src/state/service.rs

```diff
@@ -92,20 +99,19 @@ pub struct ServiceStore {
     workload_to_vips: HashMap<NetworkAddress, HashSet<NetworkAddress>>,
 
     /// Allows for lookup of services by network address, the service's xds secondary key.
-    /// this map stores the same services as `services_by_hostname` so we can mutate the services in both maps
-    /// at once when a service is updated or workload updates affect a service's endpoints.
-    by_vip: HashMap<NetworkAddress, Arc<RwLock<Service>>>,
-
-    /// Allows for lookup of services by namespaced and hostname, the service's xds primary key.
-    /// this map stores the same services as `services_by_ip` so we can mutate the services in both maps
-    /// at once when a service is updated or workload updates affect a service's endpoints.
-    by_host: HashMap<String, Vec<Arc<RwLock<Service>>>>,
+    by_vip: HashMap<NetworkAddress, Arc<Service>>,
```

Comment:
I am not sure we need the Arc? this seems to work:

```diff
diff --git a/src/cert_fetcher.rs b/src/cert_fetcher.rs
index 607d1ab..033b61e 100644
--- a/src/cert_fetcher.rs
+++ b/src/cert_fetcher.rs
@@ -17,7 +17,7 @@ use crate::config::ProxyMode;
 use crate::identity::Priority::Warmup;
 use crate::identity::{Identity, SecretManager};
 use crate::state::workload::{Protocol, Workload};
-use std::sync::{Arc, Mutex};
+use std::sync::Arc;
 use tokio::sync::mpsc;
 use tracing::{debug, error, info};
 
@@ -45,7 +45,7 @@ pub fn new(cfg: &config::Config, cert_manager: Arc<SecretManager>) -> Arc<dyn Ce
 struct CertFetcherImpl {
     proxy_mode: ProxyMode,
     local_node: Option<String>,
-    tx: Mutex<mpsc::Sender<Identity>>,
+    tx: mpsc::Sender<Identity>,
 }
 
 impl CertFetcherImpl {
@@ -72,7 +72,7 @@ impl CertFetcherImpl {
         Self {
             proxy_mode: cfg.proxy_mode.clone(),
             local_node: cfg.local_node.clone(),
-            tx: Mutex::new(tx),
+            tx,
         }
     }
 
@@ -92,8 +92,7 @@ impl CertFetcherImpl {
 impl CertFetcher for CertFetcherImpl {
     fn prefetch_cert(&self, w: &Workload) {
         if self.should_prefetch_certificate(w) {
-            let tx = self.tx.lock().unwrap();
-            if let Err(e) = tx.try_send(w.identity()) {
+            if let Err(e) = self.tx.try_send(w.identity()) {
                 info!("couldn't prefetch: {:?}", e)
             }
         }
diff --git a/src/state.rs b/src/state.rs
index 9588252..8ceca8b 100644
--- a/src/state.rs
+++ b/src/state.rs
@@ -72,7 +72,7 @@ pub struct ProxyState {
 #[derive(serde::Serialize, Debug, Clone)]
 pub struct DemandProxyState {
     #[serde(flatten)]
-    pub state: Arc<RwLock<ProxyState>>,
+    state: Arc<RwLock<ProxyState>>,
 
     /// If present, used to request on-demand updates for workloads.
     #[serde(skip_serializing)]
diff --git a/src/state/service.rs b/src/state/service.rs
index bea47c3..3b24c7e 100644
--- a/src/state/service.rs
+++ b/src/state/service.rs
@@ -99,13 +99,13 @@ pub struct ServiceStore {
     workload_to_vips: HashMap<NetworkAddress, HashSet<NetworkAddress>>,
 
     /// Allows for lookup of services by network address, the service's xds secondary key.
-    by_vip: HashMap<NetworkAddress, Arc<Service>>,
+    by_vip: HashMap<NetworkAddress, Service>,
 
     /// Allows for lookup of services by hostname, and then by namespace. XDS uses a combination
     /// of hostname and namespace as the primary key. In most cases, there will be a single
     /// service for a given hostname. However, `ServiceEntry` allows hostnames to be overridden
     /// on a per-namespace basis.
-    by_host: HashMap<String, Vec<Arc<Service>>>,
+    by_host: HashMap<String, Vec<Service>>,
 }
 
 impl ServiceStore {
@@ -231,9 +231,6 @@ impl ServiceStore {
         let vips = service.vips.clone();
         let hostname = service.hostname.clone();
 
-        // Create the Arc.
-        let service = Arc::new(service);
-
         // Map the vips to the service.
         for vip in &vips {
             self.by_vip.insert(vip.clone(), service.clone());
@@ -314,7 +311,7 @@ impl ServiceStore {
                 }
 
                 // Remove successful.
-                Some(prev.deref().clone())
+                Some(prev)
             }
         }
     }
```

Unless this is doing some expensive clone I am missing now

Reply:
I guess in get_by_host